File: packages/client-python/src/rocketride/core/runtime/ports.py
"""
Port allocation for runtime instances.

Scans from a base port upward to find the next available port.
"""

import socket

from ..constants import CONST_DEFAULT_WEB_PORT


def find_available_port(base: int = CONST_DEFAULT_WEB_PORT) -> int:
    """Find the next available TCP port starting from base.

    Checks each port with both a connect attempt (catches processes
    listening on 0.0.0.0) and a bind attempt. Scans up to 100 ports
    above base before giving up.
    """
    for offset in range(100):
        port = base + offset
        try:
            # First check if anything is already listening
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.settimeout(0.1)
                result = s.connect_ex(('127.0.0.1', port))
                if result == 0:
                    continue  # Something is listening — skip

            # Then verify we can actually bind
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.bind(('0.0.0.0', port))
                return port
        except OSError:
            continue

    raise OSError(f'No available port found in range {base}-{base + 99}')

File: packages/client-typescript/src/client/runtime/ports.ts
/**
 * Port allocation for runtime instances.
 *
 * Scans from a base port upward to find the next available port.
 */

import { createConnection, createServer } from 'net';

const DEFAULT_PORT = 5565;

function isPortInUse(port: number): Promise<boolean> {
	return new Promise((resolve) => {
		const conn = createConnection({ host: '127.0.0.1', port }, () => {
			conn.destroy();
			resolve(true);
		});
		conn.on('error', () => resolve(false));
		conn.setTimeout(100, () => {
			conn.destroy();
			resolve(false);
		});
	});
}

function canBind(port: number): Promise<boolean> {
	return new Promise((resolve) => {
		const srv = createServer();
		srv.once('error', () => resolve(false));
		srv.listen(port, '0.0.0.0', () => {
			srv.close(() => resolve(true));
		});
	});
}

export async function findAvailablePort(base: number = DEFAULT_PORT): Promise<number> {
	for (let offset = 0; offset < 100; offset++) {
		const port = base + offset;
		if (await isPortInUse(port)) continue;
		if (await canBind(port)) return port;
	}
	throw new Error(`No available port found in range ${base}-${base + 99}`);
}
